ARROW-15351: [Doc][Guide] Additional tutorial for R bindings (#12313)
apache/arrow: docs/source/developers/guide/tutorials/r_tutorial.rst

.. _r-tutorial:

***********
R tutorials
***********

Writing Bindings Walkthrough
============================

The first R package tutorial to be included in the New Contributor's
guide is a **Walkthrough** added in the **Writing Bindings**
vignette. With time we will try to include additional tutorials
directly into this guide.

This tutorial will show how to do a binding of a C++ function
`starts_with() <https://arrow.apache.org/docs/cpp/compute.html#containment-tests>`_
to the (base) R function ``startsWith()``.

To view the tutorial follow the
`Walkthrough section of the Writing Bindings article <https://arrow.apache.org/docs/r/articles/developers/bindings.html#walkthrough>`_.

R tutorial on adding a lubridate binding
========================================

> In this tutorial we will document a simple binding contribution
> to Arrow R package following the steps specified by the

Reviewer: ```suggestion
In this tutorial, we will document the contribution of a binding
to Arrow R package following the steps specified by the
```

:ref:`quick-ref-guide` section of the guide and a more detailed
:ref:`step_by_step` section. Navigate there whenever there is
some information you may find is missing here.

The binding will be added to the ``expression.R`` file in the 
R package. But you can also follow these steps in case you are
adding a binding that will live somewhere else.

.. seealso::
   
   To read more about the philosophy behind R bindings, refer to the
   `Writing Bindings article <https://arrow.apache.org/docs/r/articles/developers/bindings.html>`_.

This tutorial is different from the :ref:`step_by_step` as we
will be working on a specific case. This tutorial is not meant
as a step-by-step guide.

**Let's start!**

Set up
------

Let's set up the Arrow repository. We presume here that Git is
already installed. Otherwise please see the :ref:`set-up` section.

Once the `Apache Arrow repository <https://github.com/apache/arrow>`_

> is forked we will clone it and add the link of the main repository

Reviewer: Do we instruct people on how to fork it?  If so, can we link to that.  If not, we should explicitly tell the reader that they'll need to fork it.

Author: We have the instructions in the guide, I will add a link to it 👍

to our upstream.

.. code:: console

   $ git clone https://github.com/<your username>/arrow.git
   $ cd arrow
   $ git remote add upstream https://github.com/apache/arrow

Building R package
------------------

> Script for building R package differs depending on the Operating
> System you are using. For this reason we will only refer to

Reviewer: ```suggestion
The steps to follow for for building the R package differs depending on the operating
system you are using. For this reason we will only refer to
```

the instructions for the building process in this tutorial.

.. seealso::

   For the **introduction** to the building process refer to the
   :ref:`build-arrow-guide` section.

>    For the **instructions** on how to build R package refer to the

Reviewer: ```suggestion
   For the **instructions** on how to build the R package refer to the
```

   `R developer docs <https://arrow.apache.org/docs/r/articles/developing.html>`_.

JIRA issue
----------

In this tutorial we will be tackling a JIRA issue for implementing
a simple binding for ``mday()`` function that will match that of the
existing R function from ``lubridate``. The issue can be found on
`this link <https://issues.apache.org/jira/browse/ARROW-14816>`_.

.. figure:: R_tutorial_jira_issue.jpeg
   :scale: 40 %
   :alt: JIRA page of the issue for the R binding.

>    JIRA page of the issue for the lubridate::mday binding.

Reviewer: ```suggestion
   JIRA page of the issue for the lubridate::mday() binding.
```

**Examine the lubridate mday() function**

Going through the `lubridate documentation <https://lubridate.tidyverse.org/reference/day.html>`_
we can see that ``mday()`` takes a date object
and returns the day of the month as a numeric object.

We can run some examples in the R console to help us understand
the function better:

.. code-block:: R

   > library(lubridate)
   > mday(as.Date("2000-12-31"))
   [1] 31
   > mday(ymd(080306))
   [1] 6

**Examine the Arrow C++ day() function**

From the `compute function documentation <https://arrow.apache.org/docs/cpp/compute.html#containment-tests>`_
we can see that ``day`` is a unary function, which means that it takes
a single data input. The data input must be a ``Temporal class`` and
the returned value is an ``Integer/numeric`` type.

The ``Temporal class`` is specified as: Date types (Date32, Date64),
Time types (Time32, Time64), Timestamp, Duration, Interval.

We can call an Arrow C++ function from an R console using ``call_function``
to see how it works:

.. code-block:: R

   > call_function("day", Scalar$create(lubridate::ymd("2000-12-31")))
   Scalar
   31

We can see that lubridate and Arrow functions operate on and return

> equivalent data types. Lubridate ``mday()`` function has no additional

Reviewer: ```suggestion
equivalent data types. lubridate's ``mday()`` function has no additional
```

more research and code corrections.

.. code-block:: R

   devtools::test(filter="datetime")

   > devtools::test(filter="datetime")
   ℹ Loading arrow
   See arrow_info() for available features
   ℹ Testing arrow
   See arrow_info() for available features
   ✔ | F W S  OK | Context
   ✖ | 1     230 | dplyr-funcs-datetime [1.4s]                                                                                             
   ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   Failure (test-dplyr-funcs-datetime.R:187:3): strftime
   ``%>%`(...)` did not throw the expected error.
   Backtrace:
    1. testthat::expect_error(...) test-dplyr-funcs-datetime.R:187:2
    2. testthat:::expect_condition_matching(...)
   ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

   ══ Results ═════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════
   Duration: 1.4 s

   [ FAIL 1 | WARN 0 | SKIP 0 | PASS 230 ]

> There is a fail we get for the ``strftime`` function but looking

Reviewer: ```suggestion
There is a failure we get for the ``strftime`` function but looking
```

at the code we see is not connected to our work. We can move on and
maybe ask others if they are getting similar fail when running the tests.
It could be we only need to rebuild the library.

Check styling
-------------

As we did not change any part of the C++ code we will not
run any linters. In case we would need to check the styling

> for the C++ part we would run ``arrow/r/lint.sh`` to lint the code.

Reviewer: We still need to run `make style` in the `r` directory to style the R code

Author: Will add this, thanks for pointing it out!

    2 files changed, 19 insertions(+)

We can use ``git log`` to check the history of commits:

.. code:: console

   $ git log
   commit ed37d3a3b3eef76b696532f10562fea85f809fab (HEAD -> ARROW-14816)
   Author: Alenka Frim <frim.alenka@gmail.com>
   Date:   Fri Jan 21 09:15:31 2022 +0100

       Adding a binding and a test for mday() lubridate

   commit c5358787ee8f7b80f067292f49e5f032854041b9 (upstream/master, upstream/HEAD, master, ARROW-15346, ARROW-10643)
   Author: Krisztián Szűcs <szucs.krisztian@gmail.com>
   Date:   Thu Jan 20 09:45:59 2022 +0900

       ARROW-15372: [C++][Gandiva] Gandiva now depends on boost/crc.hpp which is missing from the trimmed boost archive
       
       See build error https://github.com/ursacomputing/crossbow/runs/4871392838?check_suite_focus=true#step:5:11762
       
       Closes #12190 from kszucs/ARROW-15372
       
       Authored-by: Krisztián Szűcs <szucs.krisztian@gmail.com>
       Signed-off-by: Sutou Kouhei <kou@clear-code.com>

If we started the branch some time ago, we may need to rebase
to upstream master to make sure there are no merge conflicts:

.. code:: console

>    $ git pull upstream master --rebase

Reviewer: Should this be `git rebase` not `git pull`?

Author: If I understand correctly `git pull --rebase` is equal to `git fetch`& `git rebase`, just shorter (and there is something about saving you trouble when rebasing on a force-pushed rebase https://git-scm.com/book/en/v2/Git-Branching-Rebasing#_rebase_rebase_work)

Reviewer: Oh awesome, learned something new today!

ARROW-14816. That’s great, now we can make the Pull Request
by clicking on **Compare & pull request**. 

.. figure:: R_tutorial_create_pr_notice.jpeg
   :scale: 60 %
   :alt: GitHub page of the Apache Arrow repository showing a notice bar
         indicating change has been made in our branch and a Pull Request
         can be created.

   Notice bar on the Apache Arrow repository.

First we need to change the Title to **ARROW-14816: [R] Implement
bindings for lubridate::mday** in order to match it with the
JIRA issue. Note a punctuation mark was added!

We will also add a description to make it clear to others what we are trying to do.

.. figure:: R_tutorial_pr_descr.jpeg
   :scale: 50 %
   :alt: GitHub page of the Pull Request showing the editor for the
         title and a description.

   Editing the title and the description of our Pull Request.

Once we click **Create pull request** our code can be reviewed as
a Pull Request in the Apache Arrow repository.

.. figure:: R_tutorial_pr.jpeg
   :scale: 50 %
   :alt: GitHub page of the Pull Request showing the title and a
         description.

   Here it is, our Pull Request!

> The Pull Request gets connected to the JIRA issue and the CI is running.

Reviewer: ```suggestion
The pull request gets connected to the JIRA issue and the CI is running.
```
